src/core/utils/ping_sender.cpp

```diff
@@ -0,0 +1,196 @@
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the ping sender module.
+ */
+
+#include "ping_sender.hpp"
+
+#if OPENTHREAD_CONFIG_PING_SENDER_ENABLE
+
+#include "common/encoding.hpp"
+#include "common/locator-getters.hpp"
+#include "common/random.hpp"
+
+namespace ot {
+namespace Utils {
+
+using Encoding::BigEndian::HostSwap32;
+
+void PingSender::Config::SetUnspecifiedToDefault(void)
+{
+    if (mSize == 0)
+    {
+        mSize = kDefaultSize;
+    }
+
+    if (mCount == 0)
+    {
+        mCount = kDefaultCount;
+    }
+
+    if (mInterval == 0)
+    {
+        mInterval = kDefaultInterval;
+    }
+}
+
+void PingSender::Config::InvokeCallback(const Response &aResponse) const
+{
+    VerifyOrExit(mCallback != nullptr);
+    mCallback(&aResponse, mCallbackContext);
+
+exit:
+    return;
+}
+
+PingSender::PingSender(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mIdentifier(0)
+    , mTimer(aInstance, PingSender::HandleTimer)
+    , mIcmpHandler(PingSender::HandleIcmpReceive, this)
+{
+    IgnoreError(Get<Ip6::Icmp>().RegisterHandler(mIcmpHandler));
+}
+
+Error PingSender::Ping(const Config &aConfig)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(!mTimer.IsRunning(), error = kErrorBusy);
+
+    mConfig = aConfig;
+    mConfig.SetUnspecifiedToDefault();
+    VerifyOrExit(mConfig.mInterval <= Timer::kMaxDelay, error = kErrorInvalidArgs);
+
+    mIdentifier++;
+    SendPing();
+
+exit:
+    return error;
+}
+
+void PingSender::Stop(void)
+{
+    mTimer.Stop();
+    mIdentifier++;
+}
+
+void PingSender::SendPing(void)
+{
+    TimeMilli        now     = TimerMilli::GetNow();
+    Message *        message = nullptr;
+    Ip6::MessageInfo messageInfo;
+
+    messageInfo.SetPeerAddr(mConfig.GetDestination());
+    messageInfo.mHopLimit          = mConfig.mHopLimit;
+    messageInfo.mAllowZeroHopLimit = mConfig.mAllowZeroHopLimit;
+
+    message = Get<Ip6::Icmp>().NewMessage(0);
+    VerifyOrExit(message != nullptr);
+
+    SuccessOrExit(message->Append(HostSwap32(now.GetValue())));
+
+    if (mConfig.mSize > message->GetLength())
+    {
+        SuccessOrExit(message->SetLength(mConfig.mSize));
+    }
+
+    SuccessOrExit(Get<Ip6::Icmp>().SendEchoRequest(*message, messageInfo, mIdentifier));
+
+#if OPENTHREAD_CONFIG_OTNS_ENABLE
+    Get<Utils::Otns>().EmitPingRequest(mConfig.GetDestination(), mConfig.mSize, now.GetValue(), mConfig.mHopLimit);
+#endif
+
+    message = nullptr;
+
+exit:
+    FreeMessage(message);
```

Comment:
We set the `message = nullptr` at the end of this method before the `exit` label (basically if everything is successful and `message` is passed to `Ip6::Icmp` we set `message` back to `nullptr`). So with this in the `exit` label, we can use `FreeMessage(message)` which frees it if it is not already `nullptr`.

We don't track or return any `error` here. We can consider tracking error, but the main use would be to log. I initially considered adding logs in this module but did not add any to keep the code size smaller. Also I am not sure if logs in this module would add much value since `Ip6::Icmp` class already logs tx of echo request and then we have logs in `MeshForwarder` about TX/RX of IPv6 message with the type and src/dst addresses.